resolwe/observers/mixins.py

```diff
@@ -0,0 +1,59 @@
+"""Mixins for Observable ViewSets."""
+import json
+
+from django.contrib.contenttypes.models import ContentType
+
+from rest_framework.decorators import action
+from rest_framework.exceptions import NotFound
+from rest_framework.response import Response
+
+from .models import Subscription
+from .protocol import CHANGE_TYPE_CREATE, CHANGE_TYPE_DELETE, CHANGE_TYPE_UPDATE
+
+
+class ObservableMixin:
+    """A Mixin to make a model ViewSet observable.
+
+    Adds the /subscribe and /unsubscribe endpoints to the list view.
+    """
+
+    def user_has_permission(self, id, user):
+        """Verify that an object exists for a given user."""
+        return self.get_queryset().filter_for_user(user).filter(pk=id).exists()
```

Comment:
Have you compared filtering orders (`.filter_for_user(user).filter(pk=id)` and `.filter(pk=id).filter_for_user(user)`? Do they produce the same database query?

Reply:
The query is similar enough, but apparently `filter_for_user` performs some subqueries in the background, so calling `filter` first should be better.